File: src/skywolf46/iui/util/ItemCreator.java
package skywolf46.iui.util;

import org.bukkit.Material;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.entity.Player;
import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.ItemFlag;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;
import skywolf46.iui.abstraction.InventoryUI;
import skywolf46.iui.abstraction.ItemProvider;

import java.util.*;

public class ItemCreator extends ItemProvider {
    private Material material;
    private int durability = 0;
    private int amount = 1;
    private CreatorMeta meta = null;

    public ItemCreator(Material material) {
        this.material = material;
    }

    public Material getMaterial() {
        return material;
    }

    public ItemCreator setMaterial(Material material) {
        this.material = material;
        return this;
    }

    public int getDurability() {
        return durability;
    }

    public ItemCreator setDurability(int durability) {
        this.durability = durability;
        return this;
    }

    public int getAmount() {
        return amount;
    }

    public ItemCreator setAmount(int amount) {
        this.amount = amount;
        return this;
    }

    public ItemStack build() {
        ItemStack item = new ItemStack(material, amount, (short) durability);
        if (meta != null) {
            ItemMeta meta = item.getItemMeta();
            this.meta.applyMeta(meta);
            item.setItemMeta(meta);
        }
        return item;
    }

    public CreatorMeta getMeta() {
        return (meta == null ? (meta = new CreatorMeta()) : meta);
    }

    @Override
    public ItemStack getItem(Player p, Inventory inv, InventoryUI ui, int slot, HashMap<String, Object> data) {
        return build();
    }

    public class CreatorMeta {
        private String display = null;
        private List<String> lore = null;
        private HashMap<Enchantment, Integer> ench = null;
        private List<ItemFlag> flags = new ArrayList<>();
        private boolean unbreakable = false;

        public ItemCreator item() {
            return ItemCreator.this;
        }

        public ItemStack build() {
            return ItemCreator.this.build();
        }

        public CreatorMeta setDisplayName(String display) {
            this.display = display;
            return this;
        }

        public CreatorMeta addLore(String lore) {
            getLore().add(lore);
            return this;
        }

        public CreatorMeta setLore(List<String> lore) {
            this.lore = lore;
            return this;
        }

        public List<String> getLore() {
            if (lore == null)
                lore = new ArrayList<>();
            return lore;
        }

        public CreatorMeta setEnchant(Enchantment ench, int lv) {
            if (this.ench == null)
                this.ench = new HashMap<>();
            if (lv == 0)
                this.ench.remove(ench);
            else
                this.ench.put(ench, lv);
            return this;
        }

        public CreatorMeta addFlags(ItemFlag... fl) {
            flags.addAll(Arrays.asList(fl));
            return this;
        }

        public CreatorMeta allFlag() {
            flags.addAll(Arrays.asList(ItemFlag.values()));
            return this;
        }

        public CreatorMeta unbreakable(boolean val) {
            this.unbreakable = val;
            return this;
        }

        public void applyMeta(ItemMeta meta) {
            if (display != null)
                meta.setDisplayName(display);
            if (lore != null)
                meta.setLore(lore);
            if (ench != null)
                for (Map.Entry<Enchantment, Integer> ench : this.ench.entrySet())
                    meta.addEnchant(ench.getKey(), ench.getValue(), true);
            if (flags.size() != 0)
                meta.addItemFlags(flags.toArray(new ItemFlag[0]));
            meta.setUnbreakable(unbreakable);
            apply(meta);
        }

        protected void apply(ItemMeta meta) {

        }

    }

    class SkullMeta extends CreatorMeta {
        @Override
        protected void apply(ItemMeta meta) {
            super.apply(meta);
        }
    }
}

File: src/main/kotlin/skywolf46/iui/kotlin/util/ItemCreator.kt
package skywolf46.iui.kotlin.util

import org.bukkit.Material
import org.bukkit.enchantments.Enchantment
import org.bukkit.entity.Player
import org.bukkit.inventory.Inventory
import org.bukkit.inventory.ItemFlag
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.ItemMeta

import kotlin.collections.HashMap

class ItemCreator(var material: Material)  {
    var durability = 0
        private set
    var amount = 1
        private set
    private var meta: CreatorMeta? = null
    fun setMaterial(material: Material): ItemCreator {
        this.material = material
        return this
    }

    fun setDurability(durability: Int): ItemCreator {
        this.durability = durability
        return this
    }

    fun setAmount(amount: Int): ItemCreator {
        this.amount = amount
        return this
    }

    fun build(): ItemStack {
        val item = ItemStack(material, amount,
            durability.toShort())
        if (meta != null) {
            val meta = item.itemMeta
            this.meta!!.applyMeta(meta)
            item.itemMeta = meta
        }
        return item
    }

    fun getMeta(): CreatorMeta {
        return if (meta == null) CreatorMeta().also { meta = it } else meta!!
    }


    open inner class CreatorMeta {
        private var display: String? = null
        private var lore: MutableList<String>? = null
        private var ench: HashMap<Enchantment, Int>? = null
        private val flags: MutableList<ItemFlag> = ArrayList()
        private var unbreakable = false
        fun item(): ItemCreator {
            return this@ItemCreator
        }

        fun build(): ItemStack {
            return this@ItemCreator.build()
        }

        fun setDisplayName(display: String?): CreatorMeta {
            this.display = display
            return this
        }

        fun addLore(lore: String): CreatorMeta {
            getLore().add(lore)
            return this
        }

        fun setLore(lore: MutableList<String>?): CreatorMeta {
            this.lore = lore
            return this
        }

        fun getLore(): MutableList<String> {
            if (lore == null) lore = ArrayList()
            return lore as MutableList<String>
        }

        fun setEnchant(ench: Enchantment, lv: Int): CreatorMeta {
            if (this.ench == null) this.ench = HashMap()
            if (lv == 0) this.ench!!.remove(ench) else this.ench!![ench] = lv
            return this
        }

        fun addFlags(vararg fl: ItemFlag): CreatorMeta {
            flags.addAll(listOf(*fl))
            return this
        }

        fun allFlag(): CreatorMeta {
            flags.addAll(listOf(*ItemFlag.values()))
            return this
        }

        fun unbreakable(`val`: Boolean): CreatorMeta {
            unbreakable = `val`
            return this
        }

        fun applyMeta(meta: ItemMeta) {
            if (display != null) meta.displayName = display
            if (lore != null) meta.lore = lore
            if (ench != null) for ((key, value) in ench!!) meta.addEnchant(
                key, value, true)
            if (flags.size != 0) meta.addItemFlags(*flags.toTypedArray())
            meta.isUnbreakable = unbreakable
            apply(meta)
        }

        protected open fun apply(meta: ItemMeta?) {}
    }

    internal inner class SkullMeta : CreatorMeta() {
        override fun apply(meta: ItemMeta?) {
            super.apply(meta)
        }
    }
}